### PR TITLE
feat: add DEX to LLM

### PR DESCRIPTION
**Pull request by cksanders**
### 📝 Description

Add DEX to LLM 
Move dex swap utils to common
Remove DEX filter on LLD

### ❓ Context

- **Impacted projects**:  
           LLD | LLM | Common
- **Linked resource(s)**: 
            https://ledgerhq.atlassian.net/browse/LIVE-5354
            https://ledgerhq.atlassian.net/browse/LIVE-5313

### ✅ Checklist

- [ ] **Test coverage** 
- [ ] **Atomic delivery** 
- [ ] **No breaking changes** 

### 📸 Demo

Screenshot 2023-06-15 at 12 21 43
Screenshot 2023-06-15 at 12 21 26

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

**Review by sarneijim — COMMENTED** [review]
I think the DEX part is missing. You need to redirect to the web-app.

File: .changeset/grumpy-dryers-breathe.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"ledger-live-desktop": patch
+"@ledgerhq/live-common": patch
```
[the remark this answers is not in the document]
Comment (cksanders) [reply .changeset/grumpy-dryers-breathe.md:3]: this is possible and done before.

File: apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx [hunk 2]
```diff
@@ -335,8 +345,59 @@ export function SwapForm({
...
+        accountId: string | undefined;
+        provider: string;
+      }) => {
+        if (
+          !walletApiPartnerList?.enabled ||
+          !walletApiPartnerList?.params?.list.includes(provider)
+        ) {
+          return accountId;
+        }
+        const account = accounts.find(a => a.id === accountId);
```
Comment (cksanders) [inline apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx:373]:
```suggestion
        const account = accounts.find(a => a.id === accountId);
        
```

File: apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx [hunk 3]
```diff
@@ -335,8 +345,59 @@ export function SwapForm({
...
+        if (!account) return accountId;
```
Comment (cksanders) [inline apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx:374]:
```suggestion
        if (!account) return accountId;
        
```

File: apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx [hunk 4]
```diff
@@ -335,8 +345,59 @@ export function SwapForm({
...
+        const parentAccount = isTokenAccount(account)
+          ? getParentAccount(account, accounts)
+          : undefined;
+        const walletApiId = accountToWalletAPIAccount(account, parentAccount)?.id;
+        return walletApiId || accountId;
+      };
+      const accountId = getAccountId({ accountId: fromAccountId, provider });
```
Comment (cksanders) [inline apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx:381]:
```suggestion
      const accountId = getAccountId({ accountId: fromAccountId, provider });
      
```

File: apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx [hunk 5]
```diff
@@ -380,16 +441,18 @@ export function SwapForm({
...
       swapTransaction.setTransaction(params.transaction);
     }
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [params]);
 
+  useEffect(() => {
+    if (params?.rate) {
+      setExchangeRate(params.rate);
+    }
+  }, [params, setExchangeRate, swapTransaction]);
```
Comment (cksanders) [inline apps/ledger-live-mobile/src/screens/Swap/Form/index.tsx:454]:
```suggestion
  }, [params, setExchangeRate]);
```

File: libs/ledger-live-common/src/exchange/swap/utils/index.test.ts [hunk 6]
```diff
@@ -346,3 +347,32 @@ describe("swap/utils/getNoticeType", function () {
...
+  it("should convert correct 1inch URL", async () => {
+    const customDappUrl = getCustomDappUrl({
+      provider: "oneinch",
+      providerURL: "/platform/1inch/#/1/unified/swap/eth/usdt?sourceTokenAmount=0.04",
+    });
+    expect(customDappUrl).toBe(
+      "https://app.1inch.io/#/1/simple/swap/eth/usdt?ledgerLive=true&sourceTokenAmount=0.04",
+    );
+  });
+  it("should not convert correct when backend send a complete url (future)", async () => {
```
Comment (cksanders) [inline libs/ledger-live-common/src/exchange/swap/utils/index.test.ts:371]:
```suggestion
  it("should not update URL when complete URL is provided", async () => {
```

File: libs/ledger-live-common/src/exchange/swap/utils/index.ts [hunk 7]
```diff
@@ -206,3 +213,64 @@ export const getNoticeType = (provider: string): { message: string; learnMore: b
...
+};
+
+// Back-end is not retriving the correct URL.
+
+// backend: /platform/paraswap/#/0xdac17f958d2ee523a2206206994597c13d831ec7-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599/21.3?network=1
+// real: https://embedded.paraswap.io?referrer=ledger2&embed=true&enableStaking=false&displayMenu=false&enableNetworkSwitch=false&network=1#/0xdac17f958d2ee523a2206206994597c13d831ec7-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599/21.3
+
+// backend: /platform/1inch/#/1/unified/swap/usdt/shib?sourceTokenAmount=24.6
+// real: https://app.1inch.io/#/1/simple/swap/usdt/shib?ledgerLive=true&sourceTokenAmount=24.6
+
```
Comment (cksanders) [inline libs/ledger-live-common/src/exchange/swap/utils/index.ts:232]:
```suggestion
/**
 * Get complete DAPP URL
 * @param provider
 * @param providerURL
 *
 * This Func is to ensure a complete DAPP URL is generated if partial & incorrect path is provided.
 *
 * Example 1:
 *  actual: /platform/paraswap/#/0xdac17f958d2ee523a2206206994597c13d831ec7-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599/21.3?network=1
 *  expected: https://embedded.paraswap.io?referrer=ledger2&embed=true&enableStaking=false&displayMenu=false&enableNetworkSwitch=false&network=1#/0xdac17f958d2ee523a2206206994597c13d831ec7-0x2260fac5e5542a773aa44fbcfedf7c193bc2c599/21.3
 *
 * Example 2:
 *  expected: /platform/1inch/#/1/unified/swap/usdt/shib?sourceTokenAmount=24.6
 *  actual: https://app.1inch.io/#/1/simple/swap/usdt/shib?ledgerLive=true&sourceTokenAmount=24.6
 */
```

File: libs/ledger-live-common/src/exchange/swap/utils/index.ts [hunk 8]
```diff
@@ -206,3 +213,64 @@ export const getNoticeType = (provider: string): { message: string; learnMore: b
...
+export const getCustomDappUrl = ({
+  provider,
+  providerURL = "",
+}: {
+  provider: string;
+  providerURL?: string;
+}) => {
+  // To support when bk apply set the complete URL
```
Comment (cksanders) [inline libs/ledger-live-common/src/exchange/swap/utils/index.ts:240]:
```suggestion
```
redundant as code is clear.

File: libs/ledger-live-common/src/exchange/swap/utils/index.ts [hunk 9]
```diff
@@ -206,3 +213,64 @@ export const getNoticeType = (provider: string): { message: string; learnMore: b
...
+  if (isValidUrl(providerURL)) {
+    return providerURL;
+  }
```
Comment (cksanders) [inline libs/ledger-live-common/src/exchange/swap/utils/index.ts:243]:
```suggestion
  }
  
```

File: libs/ledger-live-common/src/exchange/swap/utils/index.ts [hunk 10]
```diff
@@ -206,3 +213,64 @@ export const getNoticeType = (provider: string): { message: string; learnMore: b
...
+  const dappUrl =
+    provider === "paraswap"
+      ? "https://embedded.paraswap.io/?referrer=ledger2&embed=true&enableStaking=false&displayMenu=false&enableNetworkSwitch=false"
+      : "https://app.1inch.io/?ledgerLive=true";
+  const newUrl = `https://www.prefix.com/${providerURL}`;
+  const isValidNewdUrl = isValidUrl(newUrl);
```
Comment (cksanders) [inline libs/ledger-live-common/src/exchange/swap/utils/index.ts:249]:
```suggestion
  const isValidNewdUrl = isValidUrl(newUrl);
  
```

File: libs/ledger-live-common/src/exchange/swap/utils/index.ts [hunk 11]
```diff
@@ -206,3 +213,64 @@ export const getNoticeType = (provider: string): { message: string; learnMore: b
...
+      ...Object.fromEntries(new URLSearchParams(search)),
+      ...Object.fromEntries(urlSearchParams.entries()),
+      ...Object.fromEntries(searchParams),
+    };
+
+    // Providers need to use the standard structure: query + fragment
+    // https://www.rfc-editor.org/rfc/rfc3986#section-4.2
+    // 1inch is not using the standard  (fragnment + query).
+    // Refactor when the providers follow the standards.
+
```
Comment (cksanders) [inline libs/ledger-live-common/src/exchange/swap/utils/index.ts:265]:
```suggestion
/**
     * Providers should use the standard structure: query + fragment
     *
     * 1inch is currently not using the standard (fragment + query). To be refactored once providers follow the standard structure.
     * @see https://www.rfc-editor.org/rfc/rfc3986#section-4.2
     */
```

**Review by cksanders — COMMENTED** [review]
Looking good @sarneijim - just a few small changes I'd make.

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.tsx [hunk 12]
```diff
@@ -18,11 +19,7 @@ import { useDispatch, useSelector } from "react-redux";
 import { useHistory, useLocation } from "react-router-dom";
 import styled from "styled-components";
 import { setSwapKYCStatus } from "~/renderer/actions/settings";
-import {
-  getMainAccount,
-  getParentAccount,
-  isTokenAccount,
-} from "@ledgerhq/live-common/account/index";
+import { getParentAccount, isTokenAccount } from "@ledgerhq/live-common/account/index";
```
[the remark this answers is not in the document]
Comment (sarneijim) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.tsx:22]: For this kind of suggestions, without the index the import is not recognised.

[the remark this answers is not in the document]
Comment (cksanders) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.tsx:22]: we should fix the tcs config then.

**Review by cksanders — COMMENTED** [review]
approved - thanks for getting this done @sarneijim - I'd add more spacing before merging, but otherwise good work!

**Review by kallen-ledger — APPROVED** [review]

File: libs/ledger-live-common/src/exchange/swap/utils/index.test.ts [hunk 13]
```diff
@@ -346,3 +347,32 @@ describe("swap/utils/getNoticeType", function () {
     expect(result).toEqual(expectedResult);
   });
 });
+
+describe("swap/utils/getCustomDappUrl", () => {
```
[the remark this answers is not in the document]
Comment (sarneijim) [reply libs/ledger-live-common/src/exchange/swap/utils/index.test.ts:351]: yes, the url provider by back-end is not correct....:(

[the remark this answers is not in the document]
Comment (sarneijim) [reply libs/ledger-live-common/src/exchange/swap/utils/index.test.ts:351]: finally we have restore the old logic after speak with Xavier.

**Pull request merged**